services/repository/merge_upstream.go

```diff
@@ -102,10 +105,29 @@ func GetUpstreamDivergingInfo(ctx context.Context, repo *repo_model.Repository,
 		return info, nil
 	}
 
-	// TODO: if the fork repo has new commits, this call will fail:
-	// exit status 128 - fatal: Invalid symmetric difference expression aaaaaaaaaaaa...bbbbbbbbbbbb
-	// so at the moment, we are not able to handle this case, should be improved in the future
-	diff, err := git.GetDivergingCommits(ctx, repo.BaseRepo.RepoPath(), baseBranch.CommitID, forkBranch.CommitID)
+	// Add a temporary remote
+	tmpRemote := strconv.FormatInt(time.Now().UnixNano(), 10)
+	if err = gitRepo.AddRemote(tmpRemote, repo.BaseRepo.RepoPath(), false); err != nil {
+		log.Error("GetUpstreamDivergingInfo: AddRemote: %v", err)
+	}
+	defer func() {
+		if err := gitRepo.RemoveRemote(tmpRemote); err != nil {
+			log.Error("GetUpstreamDivergingInfo: RemoveRemote: %v", err)
+		}
+	}()
+
+	var remoteBranch string
+	_, remoteBranch, err = gitRepo.GetMergeBase(tmpRemote, baseBranch.CommitID, forkBranch.CommitID)
+	if err != nil {
+		log.Error("GetMergeBase: %v", err)
+	}
+
+	baseBranch.CommitID, err = git.GetFullCommitID(gitRepo.Ctx, gitRepo.Path, remoteBranch)
+	if err != nil {
+		baseBranch.CommitID = remoteBranch
+	}
+
+	diff, err := git.GetDivergingCommits(gitRepo.Ctx, gitRepo.Path, baseBranch.CommitID, forkBranch.CommitID)
```

Comment:
I think we could use another approach to detect the "DivergingCommits" without adding the temp remote:

1. call "GetUpstreamDivergingInfo" to see if fork is behind base (old logic)
2. if fails, check if the base's update time is newer (old logic)
3. if the base's update time is older, check whether fork's branch contains the base's head commit
    * if fork contains base's head commit: then fork is after base (no need to sync)
    * if not contains: base has newer commits (could try to sync)

This approach should be faster and more stable IMO.

Reply:
Hi, I've followed your advice not to use the temp remote for git rev-list and also added the integration test.

It just checks whether the message is shown. The more thorough tests leave for the merge-upstream PR.